### ISSUE TITLE
Max-over-dim is 2,000 times slower than max

### ISSUE THREAD

**Issue by andyljones** [pytorch/pytorch#36900]
## 🐛 Bug
`.max(dim=0)` is 2,000 times slower than `.max()` on a (100m,) array of floats:
```
max                       337us per
max-over-dim           585468us per
sort-over-dim           67624us per # for comparison
```

## To Reproduce
Here's a short timing script:

```python
import torch
import time
from contextlib import contextmanager

@contextmanager
def timeit(name, R=3):
    start = time.time()
    torch.cuda.synchronize()
    yield
    torch.cuda.synchronize()
    print(f'{name:20s} {1e6*(time.time() - start)/R:8.0f}us per')

xs = torch.randn(int(100e6),).cuda()

with timeit('max'):
    xs.max()
    
with timeit('max-over-dim'):
    xs.max(0)

with timeit('sort-over-dim'):
    xs.sort(0)[-1]
```

## Expected behavior
I'd expect max-over-a-dim to be at most as slow as sort-over-a-dim.

## Environment

```
Collecting environment information...
PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 440.26
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] pytorch-memlab==0.0.4
[pip] torch==1.4.0
[pip] torchfile==0.1.0
[pip] torchvision==0.5.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               10.1.243             h6bb024c_0  
[conda] mkl                       2020.0                      166  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.15           py37ha843d7b_0  
[conda] mkl_random                1.1.0            py37hd6b4f25_0  
[conda] numpy                     1.18.1           py37h4f9e942_0  
[conda] numpy-base                1.18.1           py37hde5b4d6_1  
[conda] pytorch                   1.4.0           py3.7_cuda10.1.243_cudnn7.6.3_0    pytorch
[conda] pytorch-memlab            0.0.4                    pypi_0    pypi
[conda] torchfile                 0.1.0                    pypi_0    pypi
[conda] torchvision               0.5.0                py37_cu101    pytorch
```

## Additional context
The same script on a CPU tensor produces results more like what I'd expect:
```
max                     19932us per
max-over-dim            98397us per
sort-over-dim         3968380us per
```
My total guess is that `max()` simply uses an optimized codepath that isn't available to `max(0)` since it has to return both values and indices, so this issue is more to trap anyone else googling up 'pytorch max slow' or 'pytorch min slow' than to ask for an actual fix.  

cc @ezyang @gchanan @zou3519 @VitalyFedyunin @ngimel

**Comment by ngimel** [comment]
cc @XiaobingSuper

**Comment by ngimel** [comment]
Thanks for reporting. 
You are right, max() uses a better implementation than max over a dimension, especially when the output size is small (in your case one element). The short-term fix for the max performance may be
```
ii=a.argmax(0)
maxval = a.gather(0, ii.unsqueeze(0)).squeeze(0)
```
cc @XiaobingSuper  for 5x perf difference on the cpu.

**Comment by ngimel** [comment]
I'm setting it as hi-pri, because going through argmax on cuda is a relatively easy fix that will improve the performance of a common max/min op a lot

**Comment by vadimkantorov** [comment]
For the cases where indices are not needed, one option can be adding an argument `return_indices=False` so that `torch.max(dim)` would return an empty tensor in `indices` component

**Comment by XiaobingSuper** [comment]
For CPU path, there has an fast path for **return_indices=False**, but not exposed to python frontend ,  named **min_values/max_values**, for **return_indices=True**, there has an opportunity to improve the performance of max/min of dims, I am now doing that.

**Comment by ngimel** [comment]
@vadimkantorov unless you are interested in squeezing out ~10% improvements in reduction performance, the workaround I posted above can be used to produce pretty good perf numbers. The situations where I'd expect its overhead to be higher than 10% are when the size of reduction output is on the order of the size of reduction input (i.e. you are reducing over dimension of size 2-10). If there are indeed cases where such small improvements to reduction performance can cause appreciable end-to-end speed-ups, it'd be interesting to learn about them, otherwise I think the API change to max is not warranted. 
That said, maybe it makes sense to expose max_values/min_values to python. But backwards for it is not a well-defined operation, unless you also preserve indices, which makes it beside the point.

**Comment by vadimkantorov** [comment]
> That said, maybe it makes sense to expose max_values/min_values to python. But backwards for it is not a well-defined operation, unless you also preserve indices, which makes it beside the point.

Make sense!
